Dev console

christianvogt opened the pull request:
Contributes the dev console package as a plugin extension.

*Note: there are many comments on this PR. Most have been addressed. Then any comment that requires larger changes will be taken up as separate bugs logged against the dev console component in Jira*

There are several changes to console that have been broken into small commits. Most of which could be contributed as separate PRs.

### Commits

- **add action item feature to dropdown**
![image](https://user-images.githubusercontent.com/14068621/59634283-6ae4c480-911c-11e9-9acd-977eea941e98.png)

- **add knative icon**
  - new icon for knative

- **update eslint rules**
  - adjusted several rules to avoid conflicts with tsx and existing styles

- **add perspective header to main nav**
  - new nav header indicates perspective and cluster hostname
![image](https://user-images.githubusercontent.com/14068621/59634599-2d346b80-911d-11e9-8e08-b5d4e7a1c8c8.png)

- **support redirect to perspective landing page**
  - fix for perspectives to redirect to their default landing page when going to `/`

- **update vscode settings for tsdk and debug.node.autoAttach**
   - save settings in project for those using vscode

- **update types for page extensions**
  - fixing typescript errors when adopting the new pages extension

- **change Administrator perspective icon to PF CogsIcon**
  - align icon with UX

- **support children in namespace bar**
  - dev console needs to add an application dropdown to the namespace bar. This change lets dev console render the namespace bar manually with new content.

- **support namespace redirect for any URL that supports namespaces**
  - Loosen the requirement for URLs to be contributed to a static list of namespaced pages. If the URL parses like it supports namespaces, then allow the namespace bar to update the namespace.

- **add active application redux state, action, reducer**
  - adding support for active application context in a similar manner that has been done with the active namespace
  - this is in console because there are shared screens in admin and dev perspectives that need this

- **update ownerReference types**
  - updating types to avoid errors

- **export routes#getRouteWebURL**
  - exporting functionality that's used in dev console

- **export create-secret#SSHAuthSubform**
  - exporting functionality that's used in dev console

- **add dev-console dependencies**
  - adding new dependencies to root package.json

- **support highlighting active root nav items outside of a section**

- **contribute dev-console plugin**
  - this commit contains all of the dev console package

- **enable dev-console plugin in console-app**
  - adds `dev-console` package dependency to `console-app`


fyi @rohitkrai03 @vojtechszocs @spadgett

spadgett commented on frontend/package.json:
```diff
@@ -69,8 +69,10 @@
+    "d3": "^5.9.2",
     "file-saver": "1.3.x",
     "font-awesome": "4.7.x",
+    "formik": "2.0.1-rc.1",
```
We're locking to a prerelease version?

rohitkrai03 replied:
@spadgett We are using V2 because it comes with support for custom formik hooks and it makes creating custom components wrapped around formik context much easier. They're actually at 2.0.1-rc..5 but because of a [bug](https://github.com/jaredpalmer/formik/issues/1560) related to validations i had to downgrade to rc-1. I'll keep track of it and update the version to latest as soon as there's fix.

spadgett commented on frontend/public/components/nav/nav-header.tsx:
```diff
@@ -0,0 +1,40 @@
+      <Title className="oc-nav-header__title" size="md">
+        {icon} {name}
+      </Title>
+      <Text className="oc-nav-header__description" component={TextVariants.small}>
```
Cluster hostnames can be long. I feel like the wrapping here is going to get awkward.

christianvogt replied:
I'll add this.

spadgett replied:
Thanks. We should consider using a smaller font since this name will pretty much always wrap otherwise. cc @serenamarie125

spadgett commented:
Thanks @christianvogt. Exciting to see this!

Some things I noticed clicking around:

![Topology · OKD 2019-06-14 09-29-40](https://user-images.githubusercontent.com/1167259/59512583-04934400-8e87-11e9-9ab0-cd198fb3429c.png)

![Topology · OKD 2019-06-14 09-30-50](https://user-images.githubusercontent.com/1167259/59512686-399f9680-8e87-11e9-88a5-067ef33ec713.png)

![Import from git · OKD 2019-06-14 09-32-49](https://user-images.githubusercontent.com/1167259/59512781-710e4300-8e87-11e9-96f4-e4af90968fb9.png)

![+Add · OKD 2019-06-14 09-34-50](https://user-images.githubusercontent.com/1167259/59512900-b6cb0b80-8e87-11e9-8e98-0ef403df5d7e.png)

The experience seems weird when there are no applications:

![Topology · OKD 2019-06-14 09-36-41](https://user-images.githubusercontent.com/1167259/59513032-f4c82f80-8e87-11e9-80ad-1cfe446b6783.png)

![+Add · OKD 2019-06-14 09-37-41](https://user-images.githubusercontent.com/1167259/59513096-20e3b080-8e88-11e9-8713-520891f42544.png)

Somehow I got the wrong content when clicking the topology nav item. (Not sure how I got in this state.)

![Topology · OKD 2019-06-14 09-41-47](https://user-images.githubusercontent.com/1167259/59513373-ba12c700-8e88-11e9-8305-c4e6de949bae.png)

![default · Details · OKD 2019-06-14 09-43-13](https://user-images.githubusercontent.com/1167259/59513497-f21a0a00-8e88-11e9-9839-f3a990c62c66.png)

spadgett commented:
I saw this error trying to open the "Application" dropdown in the Git form.

```
Uncaught Invariant Violation: Objects are not valid as a React child (found: object with keys {name}). If you meant to render a collection of children, use an array instead.
    in a (created by DropDownRow)
    in li (created by DropDownRow)
    in DropDownRow (created by Dropdown)
    in ul (created by Dropdown)
    in div (created by Dropdown)
    in div (created by Dropdown)
    in Dropdown (created by ResourceDropdown)
    in ResourceDropdown (created by ApplicationDropdown)
    in Unknown (created by Connect(Component))
    in Connect(Component) (created by Firehose)
    in Firehose (created by Connect(Firehose))
    in Connect(Firehose) (created by ApplicationDropdown)
    in ApplicationDropdown (created by ApplicationSelector)
    in div (created by FormGroup)
    in FormGroup (created by ApplicationSelector)
    in ApplicationSelector (created by AppSection)
    in FormSection (created by AppSection)
    in AppSection (created by GitImportForm)
    in div (created by GitImportForm)
    in form (created by Form)
    in Form (created by GitImportForm)
    in GitImportForm (created by B)
    in B (created by GitImport)
    in GitImport (created by ImportPage)
    in Unknown (created by Connect(Component))
    in Connect(Component) (created by Firehose)
    in Firehose (created by Connect(Firehose))
    in Connect(Firehose) (created by ImportPage)
    in div (created by ImportPage)
    in ImportPage (created by AsyncComponent)
    in AsyncComponent (created by Route)
    in Route (created by LazyRoute)
    in LazyRoute
    in Switch
    in div
    in div
    in section (created by PageSection)
    in PageSection
    in Unknown (created by Route)
    in Route (created by withRouter(undefined))
    in withRouter(undefined) (created by App)
    in main (created by Page)
    in div (created by Page)
    in Page (created by App)
    in App (created by Route)
    in Route
    in Router
    in Provider
    at invariant (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:261812:15)
    at throwOnInvalidObjectType (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:273590:5)
    at reconcileChildFibers (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:274364:7)
    at reconcileChildren (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:276159:28)
    at updateHostComponent (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:276620:3)
    at beginWork (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:277406:14)
    at performUnitOfWork (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:281069:12)
    at workLoop (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:281109:24)
    at HTMLUnknownElement.callCallback (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:261906:14)
    at Object.invokeGuardedCallbackDev (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:261956:16)
ws-factory.js:250 destroying websocket: /api/kubernetes/apis/config.openshift.io/v1/clusterversions?watch=true&fieldSelector=metadata.name%3Dversion
ws-factory.js:250 destroying websocket: /api/kubernetes/apis/image.openshift.io/v1/imagestreams?watch=true&resourceVersion=413787
ws-factory.js:250 destroying websocket: /api/kubernetes/apis/project.openshift.io/v1/projects?watch=true
ws-factory.js:250 destroying websocket: /api/kubernetes/apis/apps.openshift.io/v1/namespaces/viraj/deploymentconfigs?watch=true&resourceVersion=413790
ws-factory.js:250 destroying websocket: /api/kubernetes/apis/apps/v1/namespaces/viraj/deployments?watch=true&resourceVersion=413790
react-dom.development.js:17117 The above error occurred in the <a> component:
    in a (created by DropDownRow)
    in li (created by DropDownRow)
    in DropDownRow (created by Dropdown)
    in ul (created by Dropdown)
    in div (created by Dropdown)
    in div (created by Dropdown)
    in Dropdown (created by ResourceDropdown)
    in ResourceDropdown (created by ApplicationDropdown)
    in Unknown (created by Connect(Component))
    in Connect(Component) (created by Firehose)
    in Firehose (created by Connect(Firehose))
    in Connect(Firehose) (created by ApplicationDropdown)
    in ApplicationDropdown (created by ApplicationSelector)
    in div (created by FormGroup)
    in FormGroup (created by ApplicationSelector)
    in ApplicationSelector (created by AppSection)
    in FormSection (created by AppSection)
    in AppSection (created by GitImportForm)
    in div (created by GitImportForm)
    in form (created by Form)
    in Form (created by GitImportForm)
    in GitImportForm (created by B)
    in B (created by GitImport)
    in GitImport (created by ImportPage)
    in Unknown (created by Connect(Component))
    in Connect(Component) (created by Firehose)
    in Firehose (created by Connect(Firehose))
    in Connect(Firehose) (created by ImportPage)
    in div (created by ImportPage)
    in ImportPage (created by AsyncComponent)
    in AsyncComponent (created by Route)
    in Route (created by LazyRoute)
    in LazyRoute
    in Switch
    in div
    in div
    in section (created by PageSection)
    in PageSection
    in Unknown (created by Route)
    in Route (created by withRouter(undefined))
    in withRouter(undefined) (created by App)
    in main (created by Page)
    in div (created by Page)
    in Page (created by App)
    in App (created by Route)
    in Route
    in Router
    in Provider

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://fb.me/react-error-boundaries to learn more about error boundaries.
logCapturedError @ react-dom.development.js:17117
react-dom.development.js:20312 Uncaught Invariant Violation: Objects are not valid as a React child (found: object with keys {name}). If you meant to render a collection of children, use an array instead.
    in a (created by DropDownRow)
    in li (created by DropDownRow)
    in DropDownRow (created by Dropdown)
    in ul (created by Dropdown)
    in div (created by Dropdown)
    in div (created by Dropdown)
    in Dropdown (created by ResourceDropdown)
    in ResourceDropdown (created by ApplicationDropdown)
    in Unknown (created by Connect(Component))
    in Connect(Component) (created by Firehose)
    in Firehose (created by Connect(Firehose))
    in Connect(Firehose) (created by ApplicationDropdown)
    in ApplicationDropdown (created by ApplicationSelector)
    in div (created by FormGroup)
    in FormGroup (created by ApplicationSelector)
    in ApplicationSelector (created by AppSection)
    in FormSection (created by AppSection)
    in AppSection (created by GitImportForm)
    in div (created by GitImportForm)
    in form (created by Form)
    in Form (created by GitImportForm)
    in GitImportForm (created by B)
    in B (created by GitImport)
    in GitImport (created by ImportPage)
    in Unknown (created by Connect(Component))
    in Connect(Component) (created by Firehose)
    in Firehose (created by Connect(Firehose))
    in Connect(Firehose) (created by ImportPage)
    in div (created by ImportPage)
    in ImportPage (created by AsyncComponent)
    in AsyncComponent (created by Route)
    in Route (created by LazyRoute)
    in LazyRoute
    in Switch
    in div
    in div
    in section (created by PageSection)
    in PageSection
    in Unknown (created by Route)
    in Route (created by withRouter(undefined))
    in withRouter(undefined) (created by App)
    in main (created by Page)
    in div (created by Page)
    in Page (created by App)
    in App (created by Route)
    in Route
    in Router
    in Provider
    at invariant (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:261812:15)
    at throwOnInvalidObjectType (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:273590:5)
    at reconcileChildFibers (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:274364:7)
    at reconcileChildren (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:276159:28)
    at updateHostComponent (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:276620:3)
    at beginWork (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:277406:14)
    at performUnitOfWork (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:281069:12)
    at workLoop (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:281109:24)
    at renderRoot (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:281192:7)
    at performWorkOnRoot (http://localhost:9000/static/vendors~main-6b817cfa623d852528dd.js:282099:7)
```

spadgett commented:
Expand/collapse sections in forms can get confusing. We had some usability bugs on that in the past. 

* It's not clear if you edit something and collapse the section whether it still takes effect
* How do you handle invalid fields in collapsed sections?

(Not saying we need to change it in this PR, just giving feedback from past experience.)

spadgett commented:
I see now that the topology view is meant to show the "add" options as an empty state. That's super confusing imo. I thought it was a bug! We should at least add a message saying there are no workloads in your project.

christianvogt commented on frontend/public/components/_dropdown.scss:
```diff
@@ -180,7 +181,7 @@ $dropdown-link-hover-border: 1px solid #bee1f4;
 
 .co-namespace-selector {
-  max-width: 60%;
+  max-width: 100%;
```
This was 60% before because when on mobile it would shrink the project dropdown and use ellipsis to keep it and the import yaml button on a single line.
With the addition of the app selector trying to fit all on one line is not feasible.
What I could do actually is add a class to change the behavior only for when the app selector is included in this bar. I would pass in an additional className to apply for our use case.

That way we maintain the original behavior:
![image](https://user-images.githubusercontent.com/14068621/59518196-2c88a480-8e93-11e9-8eb6-2435e42c4fa8.png)

While with the app selector we get this:
![image](https://user-images.githubusercontent.com/14068621/59518222-3ca08400-8e93-11e9-9ce2-79633b8f3f0a.png)

christianvogt commented:
@spadgett the expand collapse of sections in the form is temporary while we work on a new design from UX this sprint.

spadgett commented on frontend/packages/dev-console/src/components/pipelines/PipelineRuns.tsx:
```diff
@@ -0,0 +1,37 @@
+    selected: ['Succeeded'],
+    reducer: pipelineRunFilterReducer,
+    items: [
+      { id: 'Succeeded', title: 'Complete' },
```
Any reason not to match the title to id?

christianvogt replied:
AFAIK this was done according to UX. I can followup with the dev who works on pipelines.
@serenamarie125

christianvogt replied:
Abhi confirmed this is intentional after discussing the values with UX and the tekton team

serenamarie125 replied:
yes @spadgett @christianvogt the decision was to match what's in the upstream

christianvogt commented:
@spadgett with regards to the topology view showing the empty state. That is currently changing. I'll share the design doc with you.
There is a nav bug i see where the current item isn't being highlighted because we use top level nav items and the logic is only available in sections today. I'll have to address that.

spadgett commented on frontend/packages/dev-console/src/components/pipelines/PipelineVisualizationStepList.scss:
```diff
@@ -0,0 +1,20 @@
+  list-style: none;
+  padding: 0;
+  width: 10em;
+  & > li {
```
Better to use BEM class names

christianvogt replied:
Is it really necessary when using descendent selectors?
I can make the change for consistency.

christianvogt commented on frontend/public/components/utils/cluster-name.ts:
```diff
@@ -0,0 +1,4 @@
+export function getClusterName(server = window.SERVER_FLAGS.kubeAPIServerURL): string {
+  const url = new URL(server);
+  return url.host.replace(/\./g, '-');
```
confirmed with PM and will change this to use the standard hostname instead of converting to dashes.

christianvogt commented on frontend/packages/dev-console/src/components/formik-fields/CheckboxField.tsx:
```diff
@@ -0,0 +1,24 @@
+      <Checkbox {...field} {...props} checked={field.value} aria-describedby={`${props.name}-help`}>
+        {label}
+      </Checkbox>
+      {helpText && <HelpBlock id={`${props.name}-help`}>{helpText}</HelpBlock>}
```
Unique IDs is something we need to tackle across the app and we should be generating IDs in the underlying form components and at least having a component that helps manage the ID generation such as the one used in PF4. I'd like to tackle this post merge.

christianvogt commented on frontend/packages/dev-console/src/components/EmptyState.tsx:
```diff
@@ -0,0 +1,100 @@
+            <CardHeader>Import from Git</CardHeader>
+            <CardBody>Import code from your git repository, to be built and deployed </CardBody>
+            <CardFooter>
+              <Link className="pf-c-button pf-m-secondary" to="/import">
```
Good point. I've raised your concern with UX. This is something we can add to our backlog for 4.2. We have a story to rework the empty state patterns with new UX. I've added your concerns there.

christianvogt commented on frontend/packages/dev-console/src/components/import/builder/BuilderImageCard.scss:
```diff
@@ -0,0 +1,40 @@
+  width: 136px;
+  margin: 4px;
+  align-items: center;
+  border: 0;
```
I wasn't aware that was the guidelines. We should discuss further when we setup stylelint.

Not that we adhered to any rules but i think there's value in grouping like rules together vs alphabetical. Eg colors vs layout

christianvogt commented on frontend/packages/dev-console/src/components/import/git/SourceSecretSelector.tsx:
```diff
@@ -0,0 +1,85 @@
+              type="text"
+              onChange={onInputChange}
+              value={sourceSecret}
+              aria-describedby="name-help"
```
I don't think it does in this case. Will remove. I think the use of `aria-describedby` is very inconsistent.

christianvogt commented on frontend/packages/dev-console/src/components/import/git/GitSection.tsx:
```diff
@@ -0,0 +1,68 @@
+          type="text"
+          name="git.ref"
+          label="Git Reference"
+          helpText="Optional branch, tag or commit."
```
Will do.

christianvogt commented on frontend/packages/dev-console/src/components/pipelineruns/PipelineRunRow.tsx:
```diff
@@ -0,0 +1,67 @@
+        pipelinerun.spec.trigger &&
+        pipelinerun.spec.trigger.type
+          ? pipelinerun.spec.trigger.type
+          : '-'}
```
I believe this was done because `_.get` isn't typesafe. That being said it's inconsistently being used in some places. Do we want to create a typesafe variant?

christianvogt commented on frontend/packages/dev-console/src/components/pipelines/PipelineDetails.tsx:
```diff
@@ -0,0 +1,41 @@
+              <React.Fragment key={task.name}>
+                <dt>Name: {task.name}</dt>
+                <dd>
+                  Ref:{' '}
```
non-breaking vs breaking space ¯\_(ツ)_/¯

This is the standard way in react to ensure a space gets written.

christianvogt commented on frontend/packages/dev-console/src/components/pipelines/PipelineHeader.tsx:
```diff
@@ -0,0 +1,40 @@
+import { ColHead, ListHeader } from '@console/internal/components/factory';
+
+const PipelineHeader = (props) => (
+  <ListHeader>
```
Damn. Missed the boat lol
I can look at updating our lists....

serenamarie125 replied:
I wasn't aware that had landed yet - sorry, thought Patrick was still working on those :]

christianvogt commented on frontend/packages/dev-console/src/components/import/route/RouteSection.tsx:
```diff
@@ -0,0 +1,17 @@
+      <CheckboxField
+        type="checkbbox"
+        name="route.create"
+        label="Create a route to the application"
```
We are following the UX design.
FYI @serenamarie125 @parvathyvr

christianvogt commented on frontend/packages/dev-console/src/components/pipelineruns/PipelineRunRow.tsx:
```diff
@@ -0,0 +1,67 @@
+  const menuActions = [
+    reRunPipelineRun(pipelinerun),
+    stopPipelineRun(pipelinerun),
+    Kebab.factory.Edit,
```
Will update all locations where this is not using common.

christianvogt commented on frontend/packages/dev-console/src/components/pipelineruns/PipelineRunRow.tsx:
```diff
@@ -0,0 +1,67 @@
+      </div>
+      <div className="col-lg-2 col-md-2 col-sm-4 col-xs-4">
+        <Timestamp
+          timestamp={
```
will update

spadgett commented:
Project bar wrapping is a little awkward at some widths. If nothing else, the bottom margin looks off.

![Topology · OKD 2019-06-14 13-27-02](https://user-images.githubusercontent.com/1167259/59526914-218c3f00-8ea8-11e9-93b6-536baee3ac04.png)

spadgett commented:
Details sidebar uses different font sizes in the different contexts:

![Monosnap 2019-06-14 13-28-42](https://user-images.githubusercontent.com/1167259/59526998-631cea00-8ea8-11e9-803a-1115ea6a6663.png)

spadgett commented:
"Create New Application" option has a placeholder style when it shouldn't. (Our `Dropdown` component has an extremely confusing API. We've talked about transitioning away from it.)

![Import from git · OKD 2019-06-14 13-30-49](https://user-images.githubusercontent.com/1167259/59527088-a5462b80-8ea8-11e9-8051-a3904cbd1a40.png)

spadgett commented:
The dev console label selector is consistent with other label selectors in console:

![Import from git · OKD 2019-06-14 13-35-25](https://user-images.githubusercontent.com/1167259/59527333-4d5bf480-8ea9-11e9-8e1c-d0fea2416d79.png)

vs

![apiservice-cabundle-injector · Details · OKD 2019-06-14 13-36-09](https://user-images.githubusercontent.com/1167259/59527359-61075b00-8ea9-11e9-91ed-51b5d95ad69c.png)

and

![Deploy Image · OKD 2019-06-14 13-36-53](https://user-images.githubusercontent.com/1167259/59527407-7e3c2980-8ea9-11e9-9b79-141354b5e770.png)

spadgett commented on frontend/packages/dev-console/src/components/pipelines/PipelineDetailsPage.tsx:
```diff
@@ -0,0 +1,61 @@
+
+  componentDidMount() {
+    // eslint-disable-next-line promise/catch-or-return
+    k8sGet(PipelineModel, this.props.name, this.props.namespace).then((res) => {
```
Error handling?

christianvogt replied:
I will create separate bugs for the team to address these.

spadgett commented on frontend/packages/dev-console/src/components/pipelines/PipelineDetailsPage.tsx:
```diff
@@ -0,0 +1,61 @@
+      }).then((listres) => {
+        this.setState({
+          menuActions: [
+            triggerPipeline(res, getLatestRun({ data: listres }, 'creationTimestamp'), 'pipelines'),
```
We should add RBAC checks to the menu items. For example,

https://github.com/openshift/console/blob/master/frontend/public/components/utils/kebab.tsx#L65-L78

spadgett commented:
For new code, we're moving to functional components and hooks instead of class-based components. Pretty sure you're aware and most of this was written before we bumped React versions, but I wanted to mention it.

spadgett commented:
White tooltip text on a gray background isn't legible:

![Screen Shot 2019-06-14 at 2 22 44 PM](https://user-images.githubusercontent.com/1167259/59529786-fad20680-8eaf-11e9-9c5a-a1357bcac5d1.png)

spadgett commented:
![Topology · OKD 2019-06-14 14-43-35](https://user-images.githubusercontent.com/1167259/59530980-e2afb680-8eb2-11e9-9999-3181c1b1a88a.png)

spadgett commented:
Import YAML button has different styles in developer console vs admin console:

![Monosnap 2019-06-14 14-47-19](https://user-images.githubusercontent.com/1167259/59531240-73869200-8eb3-11e9-872e-ed1bb773a275.png)

spadgett commented:
/assign

rohitkrai03 commented:
I am getting `https` as cluster name in the sidebar.

<img width="289" alt="Screenshot 2019-06-15 at 12 56 33 AM" src="https://user-images.githubusercontent.com/6041994/59533258-8b3c3580-8f08-11e9-9fed-3be41faea994.png">

christianvogt commented on frontend/public/components/nav/nav-header.tsx:
```diff
@@ -0,0 +1,49 @@
+import * as React from 'react';
+import { connect } from 'react-redux';
+import { Title, Text, TextVariants } from '@patternfly/react-core';
+import { getActivePerspective } from '@console/internal/reducers/ui';
```
My mistake. totally agree.

christianvogt replied:
fixed

christianvogt commented on frontend/package.json:
```diff
@@ -31,15 +31,15 @@
     "moduleNameMapper": {
       "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/__mocks__/fileMock.js",
-      "\\.(css|less)$": "<rootDir>/__mocks__/styleMock.js"
+      "\\.(css|less|scss)$": "<rootDir>/__mocks__/styleMock.js"
```
Let the race begin :P

vojtechszocs replied:
Done in #1743

christianvogt commented on frontend/package.json:
```diff
@@ -31,15 +31,15 @@
       "<rootDir>/node_modules/(?!lodash-es|@console)"
     ],
-    "testRegex": "/__tests__/.*\\.(ts|tsx|js|jsx)$",
+    "testRegex": "/__tests__/.*\\.spec\\.(ts|tsx|js|jsx)$",
```
If that PR gets in first then ok! But we require this change since we have date files inside our test dirs.

vojtechszocs replied:
Done in #1743

christianvogt commented on frontend/package.json:
```diff
@@ -94,6 +96,7 @@
     "react-jsonschema-form": "^1.0.4",
     "react-lightweight-tooltip": "1.x",
     "react-linkify": "^0.2.2",
+    "react-measure": "^2.2.6",
```
Hmmm what's the alternative; `red-width-hook`? I suppose I could use this in conjunction with getting the client height manually from the parent node.

vojtechszocs replied:
ping @rawagner

rawagner replied:
yes, we used `ref-width-hook` instead of `react-measure` as `width` was enough for us

vojtechszocs replied:
I'm OK with adding `react-measure` if we really need it.

christianvogt commented on frontend/public/declarations.d.ts:
```diff
@@ -41,3 +41,8 @@ declare interface Window {
 declare type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;
 declare type Diff<T, K> = Omit<T, keyof K>;
+
+// Needed to prevent errors due to using jest types instead of jasmine.
```
This particular type fix is because of the use with Protractor.

christianvogt commented:
added a new commit to fix the issue where top level nav items NOT part of a section were not correctly highlighting when the path was active:
see commit named `support highlighting active root nav items outside of a section`

christianvogt commented:
@vojtechszocs @spadgett I just realize that we need another extension to add resource list page filters. Please see one of the last commits: `update resource list page extension to include filters`

Not sure if this should add on to the resource list page extension or create another one. I still need to determine the exact typings for the filter.

spadgett commented on frontend/public/components/factory/list.tsx:
```diff
@@ -193,6 +194,13 @@ const listFilters = {
 };
 
+// add filters from resource list page contributions
+plugins.registry.getResourceListPages().forEach(p => {
```
See also #1693

@mareklibra fyi

christianvogt replied:
Thanks @spadgett. I was not aware of that pr.

serenamarie125 commented:
> Details sidebar uses different font sizes in the different contexts:

just to verify - we are sharing the content in the details panel (side bar) in the Topology & Status/Workloads/Old Project view, right ?

serenamarie125 commented:
> The dev console label selector is inconsistent with other label selectors in console:

Labels should move to the Adv Sections, but yes we should be consistent with what's in 4.0 for labels.  I know a follow on PR will be coming with Adv Options, wondering if it can be addressed then.

serenamarie125 commented:
> Expand/collapse sections in forms can get confusing. We had some usability bugs on that in the past.

@spadgett we will follow up with this, thanks for pointing it out!

serenamarie125 commented:
> I see now that the topology view is meant to show the "add" options as an empty state. That's super confusing imo. I thought it was a bug! We should at least add a message saying there are no workloads in your project.

💯 - we had placeholder text in the design.  This was what is in 3.11 - can we go with the same ? 
_Add content to your project from the catalog of web frameworks, databases, and other components. You may also deploy an existing image, create or replace resources from their YAML or JSON definitions, or select an item shared from another project._

christianvogt commented:
@serenamarie125 a bunch of Sam's comments were made before I fixed some layout issues. The font difference for example was already addressed.

spadgett commented on frontend/packages/dev-console/src/components/topology/topology-utils.ts:
```diff
@@ -0,0 +1,436 @@
+  );
+}
+
+function podWarnings(pod) {
```
We should line up with what the project status page does here. Note a lot of this looks like it was taken from old code written before pods had container statuses (pre k8s 1.0). 

Let's align with the simpler logic here:

https://github.com/openshift/console/blob/master/frontend/public/components/overview/index.tsx#L150-L178

christianvogt replied:
@spadgett is this something we can take up in a bug later? I'll make note and add it to my growing collection :)

spadgett replied:
> @spadgett is this something we can take up in a bug later? I'll make note and add it to my growing collection :)

Works for me

spadgett commented on frontend/packages/dev-console/src/components/topology/topology-utils.ts:
```diff
@@ -0,0 +1,436 @@
+
+function numContainersReadyFilter(pod) {
+  let numReady = 0;
+  _.forEach(pod.status.containerStatuses, function(status) {
```
style nit: prefer `=>`

(I'm guessing a lot of this logic came from origin-web-console)

christianvogt replied:
I can do a sweep now. Anytime we have nits, we should attempt to fix them with eslint so we do not have to keep coming back to them.

spadgett commented on frontend/packages/dev-console/src/components/topology/topology-utils.ts:
```diff
@@ -0,0 +1,436 @@
+  Succceeded: '#519149',
+  Terminating: '#002F5D',
+  Unknown: '#A18FFF',
+  'Scaled To 0': '#FFFFFF',
```
```suggestion
  'Scaled to 0': '#FFFFFF',
```

spadgett commented on frontend/packages/dev-console/src/components/topology/topology-utils.ts:
```diff
@@ -0,0 +1,436 @@
+   * get the route information from the service
+   * @param service
+   */
+  private getRoute(service: ResourceProps): ResourceProps {
```
It looks like a lot of the logic from the project status page is duplicated here. We should find a way to share code.

christianvogt replied:
Yes it was probably used as inspiration because at the time we didn't want to refactor the console code to help with a merge down the road. We can tackle the refactoring easier once in the same repo. Although I do not like having multiple components in a single file or having utilities depend on files with ui components. Some cleaner separation of utils and components would be good.

spadgett commented on frontend/packages/dev-console/src/components/topology/topology-utils.ts:
```diff
@@ -0,0 +1,436 @@
+  /**
+   * sort the deployement version
+   */
+  private sortByDeploymentVersion = (
```
Duplicated logic:

https://github.com/openshift/console/blob/master/frontend/public/components/overview/index.tsx#L267-L296

christianvogt replied:
Same reasoning as commented earlier. Duplication noted.

spadgett commented on frontend/packages/dev-console/src/components/topology/topology-utils.ts:
```diff
@@ -0,0 +1,436 @@
+        _.concat(this.resources.deploymentConfigs.data, this.resources.deployments.data),
+      );
+      // find and add the edges for a node
+      if (_.has(annotations, 'app.openshift.io/connects-to')) {
```
Does this work? I'd think it would be have to be the following due to special characters

```suggestion
      if (_.has(annotations, ['app.openshift.io/connects-to'])) {
```

christianvogt replied:
Ya it works. But good catch as this param is expected to be a path. I'll update.

christianvogt replied:
Updated.

vojtechszocs commented:
@christianvogt Please note, #1735 should address the requirement of using custom filter functions in `List` and `Table` components.

christianvogt commented on frontend/packages/console-plugin-sdk/src/typings/pages.ts:
```diff
@@ -15,26 +15,26 @@ namespace ExtensionProperties {
-    kind?: K8sResourceKindReference;
+    kind: K8sResourceKindReference;
     /** Whether the page should assign focus when loaded. */
     autoFocus?: boolean;
```
updated

christianvogt commented on frontend/packages/dev-console/src/components/topology/topology-utils.ts:
```diff
@@ -0,0 +1,436 @@
+    };
+    const configTemplate = _.get(deploymentConfig, 'spec.template');
+    _.each(this.selectorsByService, (selector, serviceName) => {
+      if (selector.matches(configTemplate)) {
```
Fyi @karthikjeeyar 
We need a bug to track this. Although we already have a story to remove duplicate code.

karthikjeeyar replied:
Sure, we can address these bugs in a separate PR.

spadgett commented:
I'm seeing no content in the pipelines page after pulling the latest:

![Pipelines · OKD 2019-06-18 13-18-45](https://user-images.githubusercontent.com/1167259/59705076-9ffc2000-91cb-11e9-9741-ce6b337b9671.png)

spadgett commented:
@christianvogt Console can run on native k8s. We should evaluate what's broken and either fix or disable those views. I suspect topology at least will fail since it requires OpenShift resources.

divyanshiGupta commented:
/approve

christianvogt commented:
> @christianvogt can we remove projects-bar from the projects page as it shows the list of projects when
> `all projects` is selected and if some other project is selected it shows the error page? It can just show the list of all projects as we have in console.

Thanks @divyanshiGupta I fixed this.

> @christianvogt Console can run on native k8s. We should evaluate what's broken and either fix or disable those views. I suspect topology at least will fail since it requires OpenShift resources.

@spadgett I've a story for us to track this request

As for the pipelines page, I added back the missing feature flag so that the pipelines nav links is not available unless tekton is installed.

spadgett commented:
These aren't merge blockers. Just mentioning so we have issues to track....

Dropdowns open behind the details panel:

![Topology · OKD 2019-06-18 16-19-02](https://user-images.githubusercontent.com/1167259/59716909-de520900-91e4-11e9-91b9-42ba8b49d6fd.png)

Styling of the panel is different between topology and project status, notably the border and shadow

![openshift-console · Details · OKD 2019-06-18 16-21-05](https://user-images.githubusercontent.com/1167259/59717012-1d805a00-91e5-11e9-97ca-1e46b6c199bc.png)

spadgett commented:
I know we're looking at an alternative to the expand/collapse component, but calling out some minor spacing issues anyway:

![Import from Git · OKD 2019-06-18 16-22-29](https://user-images.githubusercontent.com/1167259/59717125-69cb9a00-91e5-11e9-8b2c-746e3318573e.png)

vikram-raj commented:
Connection between node is not render properly.

![Screenshot from 2019-06-19 01-58-09](https://user-images.githubusercontent.com/2561818/59717512-c35bb500-9235-11e9-8896-f588263b10f3.png)

spadgett commented:
Topology view doesn't show daemon sets:

![openshift-apiserver · Details · OKD 2019-06-18 16-31-41](https://user-images.githubusercontent.com/1167259/59717792-bb285900-91e6-11e9-95bb-ac59f0b8986d.png)

spadgett commented:
I don't want to block the PR on remaining issues. We should make sure we have issues to track.

/approve

christianvogt commented:
@vikram-raj i checked your project. You connected the node to itself in the annotation :P I updated it, and it shows correctly.
@spadgett I will be creating bugs and stories to capture everything that you've outlined.

vikram-raj commented:
> @vikram-raj i checked your project. You connected the node to itself in the annotation :P I updated it, and it shows correctly.

Sorry, my bad.

christianvogt commented:
@spadgett I have added issues to track all your latest comments.
Should be good to go now.

christianvogt commented:
rebased

ppitonak commented:
@christianvogt `./build.sh` fails when I rebase

spadgett commented:
@christianvogt fyi, needs rebase again :/

christianvogt commented:
I just rebased everything again :)

christianvogt commented:
> @christianvogt `./build.sh` fails when I rebase

@ppitonak I just and it works for me

spadgett commented:
CI will fail if `build.sh` fails, so we should be good
